internal/highway/x/store/store.go

```diff
@@ -0,0 +1,116 @@
+package store
+
+import (
+	"context"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	"github.com/ipld/go-ipld-prime/storage/fsstore"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+var (
+	_ datastore.PersistentDatastore = (*Store)(nil)
+	_ datastore.Batching            = (*Store)(nil)
+)
+
+var ObjectKeySuffix = ".dsobject"
+
+// Store is a datastore.Batching implementation that stores data in memory.
+type Store struct {
+	path      string
+	dataStore *fsstore.Store
+}
+
+// New returns a new Store.
+func New(path string) *Store {
+	store := &fsstore.Store{}
+	err := store.InitDefaults("tmp")
+	if err != nil {
+		panic(err)
+	}
+
+	return &Store{
+		path:      path,
+		dataStore: store,
+	}
+}
+
+func (s *Store) Get(ctx context.Context, key datastore.Key) (value []byte, err error) {
+	return s.dataStore.Get(ctx, key.String())
+}
+
+func (s *Store) Has(ctx context.Context, key datastore.Key) (exists bool, err error) {
+	return s.dataStore.Has(ctx, key.String())
+}
+
+func (s *Store) GetSize(ctx context.Context, key datastore.Key) (size int, err error) {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (s *Store) Query(ctx context.Context, q query.Query) (query.Results, error) {
+	log.Printf("querying for %+v", q)
+
+	results := make(chan query.Result)
+
+	walkFn := func(path string, info os.FileInfo, _ error) error {
+		// remove ds path prefix
+		relPath, err := filepath.Rel(s.path, path)
+		if err == nil {
+			path = filepath.ToSlash(relPath)
+		}
+
+		if !info.IsDir() {
+			path = strings.TrimSuffix(path, ObjectKeySuffix)
+			var result query.Result
+			key := datastore.NewKey(path)
+			result.Entry.Key = key.String()
+			if !q.KeysOnly {
+				result.Entry.Value, result.Error = s.Get(ctx, key)
+			}
+			results <- result
+		}
+		return nil
+	}
+
+	go func() {
```

Comment:
I'm thinking this would be better as a deferred. Reason being it will be easier to reason in the scenario where multiple requests are invoking this method. However my thought process could be flawed here.

Reply:
my thinking is that we should close after the function returns in the same routine context, not spawn another routine where we could have many of these running at the same time under high load.

Reply:
@joshLong145 regarding the first comment, do you mean in case multiple requests are coming in? In that case, every `go func()` is dispatched in its own context, so I don't really see where this could be a problem, but maybe I'm missing something.

Reply:
I see what you mean. Will change it and see if that's better.